File: ipad-app/ShapeSnap.swift
// 손으로 그린 점들을 사각형, 원, 삼각형, 직선 같은 반듯한 도형으로 변환
import CoreGraphics
import Foundation

enum ShapeKind {
    case rectangle, square, circle, ellipse, triangle, line
}

struct SnappedShape {
    let kind: ShapeKind
    let outline: [CGPoint]
}

struct SnappedArrow {
    let from: CGPoint
    let to: CGPoint
}

enum ShapeSnap {
    private static let edgeSamples = 12
    private static let ellipseSamples = 60

    static func snap(_ points: [CGPoint]) -> SnappedShape? {
        guard points.count >= 8 else { return nil }

        var minX = CGFloat.greatestFiniteMagnitude
        var minY = CGFloat.greatestFiniteMagnitude
        var maxX = -CGFloat.greatestFiniteMagnitude
        var maxY = -CGFloat.greatestFiniteMagnitude
        for p in points {
            minX = min(minX, p.x); minY = min(minY, p.y)
            maxX = max(maxX, p.x); maxY = max(maxY, p.y)
        }
        let w = maxX - minX
        let h = maxY - minY
        let diag = hypot(w, h)
        guard diag >= 40 else { return nil }

        let first = points.first!
        let last = points.last!
        let gap = hypot(last.x - first.x, last.y - first.y)
        let closed = gap < 0.33 * max(w, h)

        // 직선 여부
        if !closed {
            let lineLen = hypot(last.x - first.x, last.y - first.y)
            var maxDev: CGFloat = 0
            for p in points { maxDev = max(maxDev, perpDistance(p, first, last)) }
            if lineLen > 0 && maxDev < 0.09 * lineLen {
                return SnappedShape(kind: .line, outline: [first, last])
            }
            return nil
        }

        let cx = (minX + maxX) / 2
        let cy = (minY + maxY) / 2
        let rx = w / 2
        let ry = h / 2
        let squareish = abs(w - h) / max(w, h) < 0.18

        // 타원 여부
        if rx > 4 && ry > 4 {
            var sum: CGFloat = 0
            var sum2: CGFloat = 0
            for p in points {
                let r = hypot((p.x - cx) / rx, (p.y - cy) / ry)
                sum += r
                sum2 += r * r
            }
            let n = CGFloat(points.count)
            let mean = sum / n
            let std = (max(0, sum2 / n - mean * mean)).squareRoot()
            if mean > 0.88 && mean < 1.12 && std < 0.18 {
                if squareish {
                    let r = (rx + ry) / 2
                    return SnappedShape(kind: .circle, outline: ellipsePoints(cx, cy, r, r))
                }
                return SnappedShape(kind: .ellipse, outline: ellipsePoints(cx, cy, rx, ry))
            }
        }

        // 코너 검사
        var corners = rdp(points, 0.06 * diag)
        // 닫힘으로 인한 시작=끝 중복 제거
        if corners.count > 1,
           hypot(corners[0].x - corners[corners.count - 1].x,
                 corners[0].y - corners[corners.count - 1].y) < 0.12 * diag {
            corners.removeLast()
        }
        // 너무 가까운 꼭짓점 병합
        var merged: [CGPoint] = []
        for c in corners {
            if let prev = merged.last {
                if hypot(c.x - prev.x, c.y - prev.y) > 0.13 * diag { merged.append(c) }
            } else {
                merged.append(c)
            }
        }

        if merged.count == 3 {
            return SnappedShape(kind: .triangle, outline: polygonPoints(merged))
        }
        if merged.count == 4 {
            var x0 = minX, y0 = minY, x1 = maxX, y1 = maxY
            var kind = ShapeKind.rectangle
            if squareish {
                let side = (w + h) / 2
                x0 = cx - side / 2; x1 = cx + side / 2
                y0 = cy - side / 2; y1 = cy + side / 2
                kind = .square
            }
            let rect = [
                CGPoint(x: x0, y: y0), CGPoint(x: x1, y: y0),
                CGPoint(x: x1, y: y1), CGPoint(x: x0, y: y1),
            ]
            return SnappedShape(kind: kind, outline: polygonPoints(rect))
        }

        return nil
    }

    static func snapArrow(_ points: [CGPoint]) -> SnappedArrow? {
        guard let bounds = metrics(points), bounds.diag >= 40 else { return nil }
        let simplified = rdp(points, 0.035 * bounds.diag)
        guard simplified.count >= 5 && simplified.count <= 8 else { return nil }

        let tail = simplified[0]
        let firstTip = simplified[1]
        let secondTip = simplified[simplified.count - 2]
        let tip = CGPoint(
            x: (firstTip.x + secondTip.x) / 2,
            y: (firstTip.y + secondTip.y) / 2
        )
        let shaftLength = distance(tail, tip)
        guard shaftLength >= 32,
              distance(firstTip, secondTip) <= max(16, shaftLength * 0.16)
        else { return nil }

        let firstWing = simplified[2]
        let secondWing = simplified.last!
        guard validArrowHead(
            tail: tail,
            tip: tip,
            firstWing: firstWing,
            secondWing: secondWing,
            shaftLength: shaftLength
        ) else { return nil }
        return SnappedArrow(from: tail, to: tip)
    }

    static func snapArrow(shaft: [CGPoint], head: [CGPoint]) -> SnappedArrow? {
        guard shaft.count >= 2, head.count >= 3 else { return nil }
        let shaftStart = shaft.first!
        let shaftEnd = shaft.last!
        let shaftLength = distance(shaftStart, shaftEnd)
        guard shaftLength >= 32 else { return nil }
        let maxDeviation = shaft.map { perpDistance($0, shaftStart, shaftEnd) }.max() ?? 0
        guard maxDeviation <= max(5, shaftLength * 0.1) else { return nil }

        let headBounds = metrics(head)
        guard let headBounds, headBounds.diag >= 16 else { return nil }
        let simplified = rdp(head, 0.045 * headBounds.diag)
        guard simplified.count == 3 else { return nil }
        let vertex = simplified[1]
        let firstWing = simplified[0]
        let secondWing = simplified[2]

        let endIsTip = distance(shaftEnd, vertex) <= distance(shaftStart, vertex)
        let tail = endIsTip ? shaftStart : shaftEnd
        let endpoint = endIsTip ? shaftEnd : shaftStart
        guard distance(endpoint, vertex) <= max(18, shaftLength * 0.14) else { return nil }
        let tip = CGPoint(x: (endpoint.x + vertex.x) / 2, y: (endpoint.y + vertex.y) / 2)
        guard validArrowHead(
            tail: tail,
            tip: tip,
            firstWing: firstWing,
            secondWing: secondWing,
            shaftLength: shaftLength
        ) else { return nil }
        return SnappedArrow(from: tail, to: tip)
    }

    static func arrow(from a: CGPoint, to b: CGPoint) -> [CGPoint] {
        let dx = b.x - a.x, dy = b.y - a.y
        let len = hypot(dx, dy)
        guard len > 0 else { return [a, b] }
        let ux = dx / len, uy = dy / len
        let head = min(max(len * 0.22, 12), 30)
        let c = cos(CGFloat.pi / 7), s = sin(CGFloat.pi / 7) // ≈25.7°
        let b1 = CGPoint(x: b.x - head * (ux * c - uy * s), y: b.y - head * (uy * c + ux * s))
        let b2 = CGPoint(x: b.x - head * (ux * c + uy * s), y: b.y - head * (uy * c - ux * s))
        return densify([a, b, b1, b, b2])
    }


    private static func densify(_ corners: [CGPoint]) -> [CGPoint] {
        guard corners.count >= 2 else { return corners }
        var out: [CGPoint] = []
        for i in 0..<(corners.count - 1) {
            out += edgePoints(corners[i], corners[i + 1])
        }
        out.append(corners[corners.count - 1])
        return out
    }

    private static func perpDistance(_ p: CGPoint, _ a: CGPoint, _ b: CGPoint) -> CGFloat {
        let dx = b.x - a.x
        let dy = b.y - a.y
        let len2 = dx * dx + dy * dy
        if len2 == 0 { return hypot(p.x - a.x, p.y - a.y) }
        var t = ((p.x - a.x) * dx + (p.y - a.y) * dy) / len2
        t = max(0, min(1, t))
        return hypot(p.x - (a.x + t * dx), p.y - (a.y + t * dy))
    }

    private static func validArrowHead(
        tail: CGPoint,
        tip: CGPoint,
        firstWing: CGPoint,
        secondWing: CGPoint,
        shaftLength: CGFloat
    ) -> Bool {
        let back = CGPoint(x: tail.x - tip.x, y: tail.y - tip.y)
        let wing1 = CGPoint(x: firstWing.x - tip.x, y: firstWing.y - tip.y)
        let wing2 = CGPoint(x: secondWing.x - tip.x, y: secondWing.y - tip.y)
        let firstLength = hypot(wing1.x, wing1.y)
        let secondLength = hypot(wing2.x, wing2.y)
        guard firstLength >= max(10, shaftLength * 0.08),
              secondLength >= max(10, shaftLength * 0.08),
              firstLength <= shaftLength * 0.52,
              secondLength <= shaftLength * 0.52
        else { return false }

        let backLength = max(hypot(back.x, back.y), 0.001)
        let cosine1 = (back.x * wing1.x + back.y * wing1.y) / (backLength * firstLength)
        let cosine2 = (back.x * wing2.x + back.y * wing2.y) / (backLength * secondLength)
        guard cosine1 > 0.3, cosine2 > 0.3, cosine1 < 0.99, cosine2 < 0.99 else { return false }
        let cross1 = back.x * wing1.y - back.y * wing1.x
        let cross2 = back.x * wing2.y - back.y * wing2.x
        return cross1 * cross2 < 0
    }

    private static func distance(_ a: CGPoint, _ b: CGPoint) -> CGFloat {
        hypot(a.x - b.x, a.y - b.y)
    }

    private static func metrics(_ points: [CGPoint]) -> (diag: CGFloat, width: CGFloat, height: CGFloat)? {
        guard let first = points.first else { return nil }
        var minX = first.x, maxX = first.x, minY = first.y, maxY = first.y
        for point in points.dropFirst() {
            minX = min(minX, point.x); maxX = max(maxX, point.x)
            minY = min(minY, point.y); maxY = max(maxY, point.y)
        }
        let width = maxX - minX
        let height = maxY - minY
        return (hypot(width, height), width, height)
    }


    private static func rdp(_ pts: [CGPoint], _ eps: CGFloat) -> [CGPoint] {
        guard pts.count >= 3 else { return pts }
        var maxD: CGFloat = 0
        var idx = 0
        for i in 1..<(pts.count - 1) {
            let d = perpDistance(pts[i], pts[0], pts[pts.count - 1])
            if d > maxD { maxD = d; idx = i }
        }
        if maxD > eps {
            let left = rdp(Array(pts[0...idx]), eps)
            let right = rdp(Array(pts[idx...]), eps)
            return Array(left.dropLast()) + right
        }
        return [pts[0], pts[pts.count - 1]]
    }

    private static func edgePoints(_ a: CGPoint, _ b: CGPoint) -> [CGPoint] {
        var out: [CGPoint] = []
        for i in 0..<edgeSamples {
            let t = CGFloat(i) / CGFloat(edgeSamples)
            out.append(CGPoint(x: a.x + (b.x - a.x) * t, y: a.y + (b.y - a.y) * t))
        }
        return out
    }

    private static func polygonPoints(_ corners: [CGPoint]) -> [CGPoint] {
        var out: [CGPoint] = []
        for i in 0..<corners.count {
            out += edgePoints(corners[i], corners[(i + 1) % corners.count])
        }
        out.append(corners[0]) // 닫기
        return out
    }

    private static func ellipsePoints(_ cx: CGFloat, _ cy: CGFloat, _ rx: CGFloat, _ ry: CGFloat) -> [CGPoint] {
        var out: [CGPoint] = []
        for i in 0...ellipseSamples {
            let t = CGFloat(i) / CGFloat(ellipseSamples) * 2 * .pi
            out.append(CGPoint(x: cx + rx * cos(t), y: cy + ry * sin(t)))
        }
        return out
    }
}

File: flutter-app/lib/canvas/shape_snap.dart
import 'dart:math';
import 'dart:ui';

/// 손그림 획을 깔끔한 도형으로 스냅 (네모·정사각형·원·타원·삼각형·직선).
/// ipad-app/ShapeSnap.swift의 Dart 이식.

enum ShapeKind { rectangle, square, circle, ellipse, triangle, line, arrow }

class SnappedShape {
  const SnappedShape(this.kind, this.outline);
  final ShapeKind kind;
  final List<Offset> outline;
}

class ShapeSnap {
  ShapeSnap._();

  static const _edgeSamples = 12;
  static const _ellipseSamples = 60;

  /// 점 목록을 도형으로 스냅. 인식 실패 시 null.
  static SnappedShape? snap(List<Offset> points) {
    if (points.length < 8) return null;

    var minX = double.infinity, minY = double.infinity;
    var maxX = -double.infinity, maxY = -double.infinity;
    for (final p in points) {
      minX = min(minX, p.dx);
      minY = min(minY, p.dy);
      maxX = max(maxX, p.dx);
      maxY = max(maxY, p.dy);
    }
    final w = maxX - minX;
    final h = maxY - minY;
    final diag = sqrt(w * w + h * h);
    if (diag < 40) return null; // 너무 작으면 도형으로 보지 않음

    final first = points.first;
    final last = points.last;
    final gap = (last - first).distance;
    final closed = gap < 0.33 * max(w, h);

    // ── 열린 획: 직선인가 ──────────────────────────────
    if (!closed) {
      final lineLen = (last - first).distance;
      var maxDev = 0.0;
      for (final p in points) {
        maxDev = max(maxDev, _perpDistance(p, first, last));
      }
      if (lineLen > 0 && maxDev < 0.09 * lineLen) {
        return SnappedShape(ShapeKind.line, [first, last]);
      }

      // ── 화살표: 긴 샤프트 + 끝에 모인 화살촉 꼭짓점들 ──
      final corners = _rdp(points, 0.045 * diag);
      if (corners.length >= 3 && corners.length <= 6) {
        final start = corners[0];
        final tip = corners[1];
        final shaft = (tip - start).distance;
        final headDists = [for (final c in corners.skip(2)) (c - tip).distance];
        final headOk = headDists.every((d) => d <= 0.45 * shaft);
        final headBigEnough =
            headDists.any((d) => d >= max(0.1 * shaft, 12)); // 우발적 꺾임 제외
        if (shaft >= 0.5 * diag && headOk && headBigEnough) {
          return SnappedShape(ShapeKind.arrow, arrow(start, tip));
        }
      }
      return null;
    }

    final cx = (minX + maxX) / 2;
    final cy = (minY + maxY) / 2;
    final rx = w / 2;
    final ry = h / 2;
    final squareish = (w - h).abs() / max(w, h) < 0.18;

    // ── 타원 / 원 검사 ────────────────────────────────
    // 정규화 반경 평균은 원·타원이 ≈1.0, 사각형이 ≈1.15. mean으로 사각형을
    // 걸러내고 std(떨림)는 넉넉히 둬서 손으로 대충 그린 원도 잡는다.
    if (rx > 4 && ry > 4) {
      var sum = 0.0, sum2 = 0.0;
      for (final p in points) {
        final nx = (p.dx - cx) / rx;
        final ny = (p.dy - cy) / ry;
        final r = sqrt(nx * nx + ny * ny);
        sum += r;
        sum2 += r * r;
      }
      final n = points.length.toDouble();
      final mean = sum / n;
      final std = sqrt(max(0, sum2 / n - mean * mean));
      if (mean > 0.88 && mean < 1.12 && std < 0.18) {
        if (squareish) {
          final r = (rx + ry) / 2;
          return SnappedShape(ShapeKind.circle, _ellipsePoints(cx, cy, r, r));
        }
        return SnappedShape(ShapeKind.ellipse, _ellipsePoints(cx, cy, rx, ry));
      }
    }

    // ── 코너 검사 (RDP) ───────────────────────────────
    var corners = _rdp(points, 0.06 * diag);
    // 닫힘으로 인한 시작=끝 중복 제거
    if (corners.length > 1 &&
        (corners.first - corners.last).distance < 0.12 * diag) {
      corners = corners.sublist(0, corners.length - 1);
    }
    // 너무 가까운 꼭짓점 병합
    final merged = <Offset>[];
    for (final c in corners) {
      if (merged.isEmpty || (c - merged.last).distance > 0.13 * diag) {
        merged.add(c);
      }
    }

    if (merged.length == 3) {
      return SnappedShape(ShapeKind.triangle, _polygonPoints(merged));
    }
    if (merged.length == 4) {
      // 직사각형 → 축 정렬 bbox. 정사각형에 가까우면 정사각형.
      var x0 = minX, y0 = minY, x1 = maxX, y1 = maxY;
      var kind = ShapeKind.rectangle;
      if (squareish) {
        final side = (w + h) / 2;
        x0 = cx - side / 2;
        x1 = cx + side / 2;
        y0 = cy - side / 2;
        y1 = cy + side / 2;
        kind = ShapeKind.square;
      }
      final rect = [
        Offset(x0, y0),
        Offset(x1, y0),
        Offset(x1, y1),
        Offset(x0, y1),
      ];
      return SnappedShape(kind, _polygonPoints(rect));
    }

    return null; // 인식 실패 → 원래 자유곡선을 그대로 둔다
  }

  /// 직선 화살표 아웃라인(샤프트 + 화살촉). a→b 방향, 단일 폴리라인.
  static List<Offset> arrow(Offset a, Offset b) {
    final d = b - a;
    final len = d.distance;
    if (len == 0) return [a, b];
    final ux = d.dx / len, uy = d.dy / len;
    final head = (len * 0.22).clamp(12.0, 30.0);
    final c = cos(pi / 7), s = sin(pi / 7); // ≈25.7°
    // 끝점에서 ±각도로 뻗은 두 갈래. a→b→b1→b→b2 한 획으로 그린다.
    final b1 = Offset(
        b.dx - head * (ux * c - uy * s), b.dy - head * (uy * c + ux * s));
    final b2 = Offset(
        b.dx - head * (ux * c + uy * s), b.dy - head * (uy * c - ux * s));
    return _densify([a, b, b1, b, b2]);
  }

  // ── 기하 헬퍼 ──────────────────────────────────────

  static double _perpDistance(Offset p, Offset a, Offset b) {
    final dx = b.dx - a.dx, dy = b.dy - a.dy;
    final len2 = dx * dx + dy * dy;
    if (len2 == 0) return (p - a).distance;
    var t = ((p.dx - a.dx) * dx + (p.dy - a.dy) * dy) / len2;
    t = t.clamp(0.0, 1.0);
    return (p - Offset(a.dx + t * dx, a.dy + t * dy)).distance;
  }

  /// Ramer–Douglas–Peucker 단순화
  static List<Offset> _rdp(List<Offset> pts, double eps) {
    if (pts.length < 3) return List.of(pts);
    var maxD = 0.0;
    var idx = 0;
    for (var i = 1; i < pts.length - 1; i++) {
      final d = _perpDistance(pts[i], pts.first, pts.last);
      if (d > maxD) {
        maxD = d;
        idx = i;
      }
    }
    if (maxD > eps) {
      final left = _rdp(pts.sublist(0, idx + 1), eps);
      final right = _rdp(pts.sublist(idx), eps);
      return [...left.sublist(0, left.length - 1), ...right];
    }
    return [pts.first, pts.last];
  }

  static List<Offset> _edgePoints(Offset a, Offset b) => [
        for (var i = 0; i < _edgeSamples; i++)
          Offset.lerp(a, b, i / _edgeSamples)!,
      ];

  static List<Offset> _polygonPoints(List<Offset> corners) {
    final out = <Offset>[];
    for (var i = 0; i < corners.length; i++) {
      out.addAll(_edgePoints(corners[i], corners[(i + 1) % corners.length]));
    }
    out.add(corners.first); // 닫기
    return out;
  }

  /// 꼭짓점 사이를 촘촘한 점으로 채운 열린 폴리라인.
  static List<Offset> _densify(List<Offset> corners) {
    if (corners.length < 2) return corners;
    final out = <Offset>[];
    for (var i = 0; i < corners.length - 1; i++) {
      out.addAll(_edgePoints(corners[i], corners[i + 1]));
    }
    out.add(corners.last);
    return out;
  }

  static List<Offset> _ellipsePoints(
      double cx, double cy, double rx, double ry) {
    return [
      for (var i = 0; i <= _ellipseSamples; i++)
        Offset(cx + rx * cos(i / _ellipseSamples * 2 * pi),
            cy + ry * sin(i / _ellipseSamples * 2 * pi)),
    ];
  }
}
